interfaces/builtin/docker.go

```diff
@@ -0,0 +1,642 @@
+	switch securitySystem {
+	case interfaces.SecurityAppArmor,
+		interfaces.SecurityDBus,
+		interfaces.SecurityMount,
+		interfaces.SecuritySecComp,
+		interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	daemonPrivileged, _ := plug.Attrs["daemon-privileged"].(bool)
+
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		snippet := []byte(dockerConnectedPlugAppArmor)
+		if daemonPrivileged {
+			snippet = append(snippet, daemonPrivilegedAppArmor...)
+		}
+		return snippet, nil
+	case interfaces.SecuritySecComp:
+		snippet := []byte(dockerConnectedPlugSecComp)
+		if daemonPrivileged {
+			snippet = append(snippet, daemonPrivilegedSecComp...)
+		}
+		return snippet, nil
+	case interfaces.SecurityDBus,
+		interfaces.SecurityMount,
+		interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		return []byte(dockerPermanentSlotAppArmor), nil
+	case interfaces.SecuritySecComp:
+		return []byte(dockerPermanentSlotSecComp), nil
+	case interfaces.SecurityDBus,
+		interfaces.SecurityMount,
+		interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	// The docker socket is a named socket and therefore mediated by
+	// AppArmor file rules. As such, there is no additional ConnectedSlot
+	// policy to add.
+	switch securitySystem {
+	case interfaces.SecurityAppArmor,
+		interfaces.SecurityDBus,
+		interfaces.SecurityMount,
+		interfaces.SecuritySecComp,
+		interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface.Name()))
+	}
+
+	// It's fine if daemon-privileged isn't specified, but if it is,
+	// it needs to be bool and it can only by used with the docker project
+	// and Canonical
+	if v, ok := plug.Attrs["daemon-privileged"]; ok {
+		if _, ok = v.(bool); !ok {
+			return fmt.Errorf("docker plug requires bool with 'daemon-privileged'")
+		}
+		snapName := plug.Snap.Name()
+		devName := plug.Snap.Developer
+		if snapName != "docker" || (devName != "canonical" && devName != "docker") {
+			return fmt.Errorf("daemon-privileged attribute is reserved for the Docker project and Canonical")
```

Comment:
Can we please reword this as:

`"daemon-privileged attribute is reserved for the upstream docker project"`